eng/common/scripts/Git-Repo-Sync.ps1

```diff
@@ -0,0 +1,119 @@
+ #!/usr/bin/env pwsh -c
+
+<#
+.DESCRIPTION
+Sync all files from SourceRepo to TargetRepo directlly.
+
+.PARAMETER SourceRepo
+The GitHub repository that needs to be referenced.
+
+.PARAMETER SourceBranch
+The branch of Source GitHub repository that needs to be referenced.
+
+.PARAMETER TargetRepo
+The GitHub repository will synced.
+
+.PARAMETER TargetBranch
+The branch of Target GitHub repository will synced.
+
+.PARAMETER AuthToken
+A personal access token
+
+.PARAMETER Rebase
+Keep the commit record when syning.
+#>
+[CmdletBinding(SupportsShouldProcess = $true)]
+param(
+  [Parameter(Mandatory = $true)]
+  [string]$SourceRepo,
+
+  [Parameter(Mandatory = $false)]
+  [string]$SourceBranch,
+
+  [Parameter(Mandatory = $true)]
+  [string]$TargetRepo,
+
+  [Parameter(Mandatory = $false)]
+  [string]$TargetBranch,
+
+  [Parameter(Mandatory = $true)]
+  [string]$AuthToken,
+
+  [Parameter(Mandatory = $false)]
+  [string]$Rebase,
+  
+  $user="azure-sdk",
+  $email="azuresdk@microsoft.com"
+)
+
+. (Join-Path $PSScriptRoot common.ps1)
+
+Set-PsDebug -Trace 1
+if (-not (Test-Path $SourceRepo)) {
+  New-Item -Path $SourceRepo -ItemType Directory -Force
+  Set-Location $SourceRepo
+  git init
+  git remote add Source "https://$($AuthToken)@github.com/$($SourceRepo).git"
+} else {
+  Set-Location $SourceRepo
+}
+
+# Check the default branch
+if (!$SourceBranch) {
+  $defaultBranch = (git remote show Source | Out-String) -replace "(?ms).*HEAD branch: (\w+).*", '$1'
+  Write-Host "No source branch. Fetch default branch $defaultBranch."
+  $SourceBranch = $defaultBranch
+}
+
+git fetch --no-tags Source $SourceBranch
+if ($LASTEXITCODE -ne 0) {
+  Write-Host "#`#vso[task.logissue type=error]Failed to fetch $($SourceRepo):$($SourceBranch)"
+  exit 1
+}
+
+git checkout ${SourceBranch}
+
+Function FailOnError([string]$ErrorMessage, $CleanUpScripts = 0) {
+  if ($LASTEXITCODE -ne 0) {
+    Write-Host "#`#vso[task.logissue type=error]$ErrorMessage"
+    if ($CleanUpScripts -ne 0) { Invoke-Command $CleanUpScripts }
+    exit 1
+  }
+}
+
+try {
+  git remote add Target "https://$($AuthToken)@github.com/$($TargetRepo).git"
+
+  $defaultBranch = (git remote show Target | Out-String) -replace "(?ms).*HEAD branch: (\w+).*", '$1'
+  if (!$SourceBranch) {
+    $SourceBranch = $defaultBranch
+  }
+  if (!$TargetBranch) {
+    $TargetBranch = $defaultBranch
+  }
+
+  if (-not $($Rebase)) {
+    git checkout -B target_branch $($SourceBranch)
+    git push --force Target "target_branch:refs/heads/$($TargetBranch)"
```

Comment:
Why is this `--force`? We should avoid that in most cases.

Reply:
In description i had this question when we push meet conflict. What we should do ? Currently I added a parameter PushArgs that can manually confirm the push operation

Reply:
If there is a conflict we probably should fail and not force push. Force pushing can be a destructive action and might causes unexpected issues. Also most of these branches we are syncing are protected and will block a force push.

Reply:
I add PushArgs parameter that will not force push with empty.  Is that need this parameter ?

Reply:
delete PushArgs parameter